### PR TITLE
[feat] Undo

### PR DESCRIPTION
**Pull request by aateg**
The undo feature makes it possible to undo tricks.
For more information about this feature check the issue #24 

This features impacts the scoreboard and also the tricks showed on the game screen, the first because when you undo the trick the points earned need to be removed, and second because the played cards shouldn't appear on the board since they was replaced on the players hands. Just solved both problems.

Didn't created tests yet because I would like to receive feedbacks first.

### CODE REVIEW

**Comment by aateg** [comment]
Just for your knowledge. I intend to squash the commits before merge.

File: src/main/java/br/com/sbk/sbking/core/Deal.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Deal.java:299]: Yes. That's why the undoScore method is inside that condition. If the number of played cards is greater than the number of a complete trick, it means that Deal.tricks has at least two tricks. Line 297 remove the current trick, therefore the line 299 uses the second to last trick.

File: src/main/java/br/com/sbk/sbking/core/Score.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Score.java:92]: I saw that methods, but I can't use them to subtract one point from the winner of the last trick. That's why I created other method

File: src/main/java/br/com/sbk/sbking/networking/server/ClientGameSocket.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/networking/server/ClientGameSocket.java:23]: It's not the first time that happens. I just added one if else statement. Need to solve that problem with my IDE

File: src/main/java/br/com/sbk/sbking/core/Deal.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Deal.java:296]: The player is allowed to call for undo when it is not his turn?

Comment (rulojuka) [reply src/main/java/br/com/sbk/sbking/core/Deal.java:296]: Yes, it should be :)
In a normal situation, it will call UNDO just after playing (a wrong card)

File: src/main/java/br/com/sbk/sbking/core/Score.java
Comment (rulojuka) [reply src/main/java/br/com/sbk/sbking/core/Score.java:92]: You should create other methods. The idea is to check how they work. Points are not given +1 or -1, it is based on the ruleset.

File: src/main/java/br/com/sbk/sbking/core/Deal.java
Comment (rulojuka) [reply src/main/java/br/com/sbk/sbking/core/Deal.java:299]: Ah, nice.

File: src/main/java/br/com/sbk/sbking/core/Board.java
Comment (rulojuka) [inline src/main/java/br/com/sbk/sbking/core/Board.java:79]: I think it is giveBack... with an uppercase B. (EDIT: in fact, it is giveBackCard**s**ToHand**s**)
Also, you probably only need to sort after the `for`. There will be another `for` in `Direction.values`, but that number of elements will be constant. The current one increases with the size of the map. (EDIT: The `for` you wrote is constant too (max 2*trick_size) cards. Ignore what I said.)
I know, I just told you to not optimize these things, but this one is really easy, hahaha :laughing:

File: src/main/java/br/com/sbk/sbking/core/Score.java
Comment (rulojuka) [inline src/main/java/br/com/sbk/sbking/core/Score.java:44]: I think those private methods are small enough (and not reused enough) to be inlined. But that is your call.

**Review by rulojuka — COMMENTED** [review]
Looks good to me (LGTM). Nice job! :+1: 
We just need to finish the tests to assert the methods are doing what we think they are doing :)

File: src/main/java/br/com/sbk/sbking/core/Score.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Score.java]: I accepted your suggestion.

File: src/main/java/br/com/sbk/sbking/networking/server/ClientGameSocket.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/networking/server/ClientGameSocket.java]: problem solved with my IDE

File: src/main/java/br/com/sbk/sbking/gui/elements/UndoElement.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/gui/elements/UndoElement.java]: I removed that comment.

File: src/main/java/br/com/sbk/sbking/core/Trick.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Trick.java]: I don't think this undo feature should involve a refactoring. Maybe in another PR.

File: src/main/java/br/com/sbk/sbking/core/Trick.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Trick.java]: solved.

File: src/main/java/br/com/sbk/sbking/core/Deal.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Deal.java]: This method is not necessary anymore.

File: src/main/java/br/com/sbk/sbking/core/Board.java
Comment (aateg) [reply src/main/java/br/com/sbk/sbking/core/Board.java]: I changed to putCardInHand

**Comment by rulojuka** [comment]
The code seems ok but I found a bug. I sent it on  Discord.

**Review by rulojuka — APPROVED** [review]

**Review by rulojuka — CHANGES_REQUESTED** [review]
Oops, I was trying to tell GitHub that the request changes were made, but the PR was not approved yet.

**Pull request merged**
